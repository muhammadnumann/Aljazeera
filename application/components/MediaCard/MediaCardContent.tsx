import { MouseEventHandler } from 'react'

interface InputProps {
  onClick?: MouseEventHandler | undefined
  content: string
  slug: string
  truncateContentLength: number
  className: string
}

const MediaCardContent = (props: InputProps) => {
  return (
    <div onClick={props.onClick} className={props.className}>
      <span className="text-gray-800 dark:text-grayDark-800 mb-1 block">
        {props.slug}
      </span>
      <span className="text-gray-700 dark:text-grayDark-700">
        {props.content.substring(0, props.truncateContentLength)}
      </span>
    </div>
  )
}

MediaCardContent.defaultProps = {
  truncateContentLength: 30,
}

export default MediaCardContent
