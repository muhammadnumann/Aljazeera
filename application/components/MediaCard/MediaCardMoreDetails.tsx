import {
  MoreVertical,
  DeleteOutline,
  DownloadFromCloud,
} from 'components/Icons'
import { Menu, Transition } from '@headlessui/react'
import { Fragment, useRef, useState } from 'react'
import ModalDialog from 'components/ModalDialog'
import { item as ItemApi } from '@vidispine/vdt-api'
import { ItemType } from '@vidispine/types'
import ModalProgressDialog from 'components/ModalProgressDialog'
import { getDownloadLink } from 'utils/ItemUtils'

interface InputProps {
  content: string
  itemType: ItemType
  onRemoveItemClick: () => void
}

const CHUNK_SIZE = 26214400

const MediaCardMoreDetails = (props: InputProps) => {
  const [currentChunk, setCurrentChunk] = useState(0)
  const [totalChunks, setTotalChunks] = useState(0)
  const [showProgress, setShowProgress] = useState(false)
  const isCanceled = useRef(false)

  let currentChunkIndex = 0

  const downloadChunks = async (
    url: string,
    contentLength: number,
    index: number,
  ) => {
    const headers = new Headers()
    headers.append(
      'Range',
      `bytes=${index * CHUNK_SIZE}-${index * CHUNK_SIZE + CHUNK_SIZE - 1}`,
    )

    let response = await fetch(url, { headers })
    const roundedUpTotalChunks = Math.ceil(contentLength / CHUNK_SIZE)
    setCurrentChunk(++currentChunkIndex)
    setTotalChunks(roundedUpTotalChunks)

    return response.blob()
  }

  const multiThreadedDownload = async (item: ItemType) => {
    const [url, mime, size, filename, originalFilename] = await getDownloadLink(
      item.id!,
    )
    setShowProgress(true)

    let totalIndexes = Math.ceil(size / CHUNK_SIZE)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let promises: any = []

    for (let index = 0; index < totalIndexes; index++) {
      if (isCanceled.current === true) {
        break
      }

      promises.push(await downloadChunks(url, size, index))
    }

    if (isCanceled.current === true) {
      isCanceled.current = false
      setShowProgress(false)
      setCurrentChunk(0)
      setTotalChunks(0)

      return
    }

    const arrayOfPromises = await Promise.all(promises)

    setShowProgress(false)
    setCurrentChunk(0)
    setTotalChunks(0)

    const data = arrayOfPromises.reduce(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (previousItem, currentItem): any => {
        return new Blob([previousItem, currentItem], { type: mime })
      },
    )

    const blobUrl = URL.createObjectURL(data)
    const a = document.createElement('a')
    a.download = originalFilename || filename
    a.href = blobUrl
    a.click()
    URL.revokeObjectURL(blobUrl)
  }

  const [isModalOpen, setIsModalOpen] = useState(false)

  const closeModal = () => {
    setIsModalOpen(false)
  }

  const openModal = () => {
    setIsModalOpen(true)
  }

  const onRemoveItem = async (value: string) => {
    const response = await ItemApi.removeItem({ itemId: value })
    setIsModalOpen(false)
    if (response.status === 200) {
      props.onRemoveItemClick()
    }
  }

  return (
    <>
      <div className="absolute top-16 w-64 text-right">
        <Menu as="div" className="relative inline-block text-left">
          <div>
            <Menu.Button className="inline-flex w-full justify-center rounded-md px-2 py-1 text-sm font-medium text-gray-800 dark:text-white focus:outline-none focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75">
              <MoreVertical
                className="h-5 w-5 fill-gray-800 dark:fill-white"
                aria-hidden="true"
              />
            </Menu.Button>
          </div>
          <Transition
            as={Fragment}
            enter="transition ease-out duration-100"
            enterFrom="transform opacity-0 scale-95"
            enterTo="transform opacity-100 scale-100"
            leave="transition ease-in duration-75"
            leaveFrom="transform opacity-100 scale-100"
            leaveTo="transform opacity-0 scale-95"
          >
            <Menu.Items className="absolute right-0 w-56 z-50 origin-top-right divide-y divide-gray-100 rounded-md  bg-gray-50 dark:bg-gray-700 shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none">
              <div className="px-1 py-1 ">
                <Menu.Item>
                  {({ active }) => (
                    <button
                      className={`${
                        active
                          ? 'bg-gray-200 dark:bg-gray-600 text-black-600 dark:text-white hover:text-gray-900'
                          : 'text-black-900 dark:text-white'
                      } group flex w-full items-center rounded-md px-2 py-2 text-sm`}
                      onClick={() => multiThreadedDownload(props.itemType)}
                    >
                      <DownloadFromCloud
                        className="mr-2 h-5 w-5 fill-black-600 dark:fill-white hover:fill-gray-900"
                        aria-hidden="true"
                      />
                      Download Item
                    </button>
                  )}
                </Menu.Item>
              </div>
              <div className="px-1 py-1 ">
                <Menu.Item>
                  {({ active }) => (
                    <button
                      className={`${
                        active
                          ? 'bg-gray-200 dark:bg-gray-600 text-black-600 dark:text-white hover:text-gray-900'
                          : 'text-black-900 dark:text-white'
                      } group flex w-full items-center rounded-md px-2 py-2 text-sm`}
                      onClick={openModal}
                    >
                      <DeleteOutline
                        className="mr-2 h-5 w-5 fill-black-600 dark:fill-white hover:fill-gray-900"
                        aria-hidden="true"
                      />
                      Remove Item
                    </button>
                  )}
                </Menu.Item>
              </div>
            </Menu.Items>
          </Transition>
        </Menu>
      </div>

      <ModalDialog
        show={isModalOpen}
        title="Remove Item"
        description={`Are you sure you want to remove the item: "${props.content}" ? `}
        ok="Yes"
        cancel="No"
        okDisabled={false}
        onClick={() => onRemoveItem(props.itemType.id ?? '')}
        onClose={closeModal}
        okClassName="inline-flex justify-center rounded-md border border-transparent bg-red-600 px-4 py-2 text-sm font-medium text-black-900 hover:bg-red-700 focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2"
        cancelClassName="ml-2 inline-flex justify-center rounded-md border border-transparent bg-gray-200 px-4 py-2 text-sm font-medium text-black-900 hover:bg-gray-300 focus:outline-none focus-visible:ring-2 focus-visible:ring-blue-500 focus-visible:ring-offset-2"
      />

      <ModalProgressDialog
        currentProgress={currentChunk}
        total={totalChunks}
        show={showProgress}
        title={`Downloading ${props.itemType.id} ...`}
        onCancel={() => (isCanceled.current = true)}
      />
    </>
  )
}

export default MediaCardMoreDetails
