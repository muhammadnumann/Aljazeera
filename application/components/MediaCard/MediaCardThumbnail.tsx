import React, {
  MouseEvent,
  MouseEventHandler,
  useContext,
  useRef,
  useState,
} from 'react'
import AppContext from 'utils/contexts/AppContext'
import { ThumbnailType } from '../../types'

interface InputProps {
  thumbnail: string
  itemId?: string
  onClick?: MouseEventHandler
  spriteImage?: HTMLImageElement
  thumbnailArray?: Array<ThumbnailType>
}

const MediaCardThumbnail = (props: InputProps) => {
  const context = useContext(AppContext)
  const [scrubPosition, setScrubPosition] = useState(0)
  const [showScrub, setShowScrub] = useState(false)
  const [srcThumbnail, setSrcThumbnail] = useState(
    `${context.serverUrl}${props.thumbnail}`,
  )
  const [x, setX] = useState(0)
  const [y, setY] = useState(0)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const divThumbnail = useRef<any>()

  const { itemId, spriteImage, thumbnailArray } = props

  const getOffsetLeft = (element: HTMLElement) => {
    try {
      return Math.round(element.getBoundingClientRect().left)
    } catch (error) {
      return 0
    }
  }

  const getMouseElementPosition = (e: MouseEvent<HTMLImageElement>) => {
    const mouseElementPosition = e.clientX - getOffsetLeft(e.currentTarget)
    const elementWidth = e.currentTarget!.offsetWidth
    if (mouseElementPosition < 0) {
      return 0
    }
    if (mouseElementPosition > elementWidth) {
      return elementWidth
    }
    return mouseElementPosition
  }

  const onMouseMove = (e: MouseEvent<HTMLImageElement>) => {
    if (spriteImage && spriteImage.complete) {
      console.info('spriteImage.complete')
      const mouseElementPosition = getMouseElementPosition(e)
      /*
       *   Calculate the thumbnail based on the mouse position
       *   index = (mouse position / ( element width / number of thumbnails))
       */
      // @ts-ignore
      const index = Math.floor(
        mouseElementPosition /
          ((divThumbnail.current ? divThumbnail.current.offsetWidth : 0) /
            thumbnailArray!.length || 1),
      )
      if (!!thumbnailArray && index >= 0 && index < thumbnailArray.length) {
        setX(thumbnailArray[index].x)
        setY(thumbnailArray[index].y)
      }
      setScrubPosition(mouseElementPosition - 1)
    }
  }

  const handleShowScrub = (value: boolean) => () => {
    if (spriteImage && spriteImage.complete) {
      setSrcThumbnail(spriteImage.src)
      setShowScrub(value)
    }
  }

  return (
    <div
      className="flex flex-col items-center justify-center cursor-pointer"
      onMouseEnter={handleShowScrub(true)}
      onMouseLeave={handleShowScrub(false)}
      onMouseMove={onMouseMove}
      onClick={props.onClick}
    >
      <div className="flex flex-col items-center justify-center relative">
        <div
          ref={divThumbnail}
          className={
            'rounded-t-lg w-[256px] h-[140px] overflow-hidden relative'
          }
        >
          <img
            alt={`${itemId} thumbnail`}
            className={`rounded-t-lg max-w-none absolute ${
              spriteImage?.complete ? 'invisible' : 'visible'
            }`}
            src={`${context.serverUrl}${props.thumbnail}`}
          />
          <img
            alt={`${itemId} thumbnail`}
            className={`rounded-t-lg max-w-none absolute ${
              spriteImage?.complete ? 'visible' : 'invisible'
            }`}
            style={{
              top: `-${y}px`,
              left: `-${x}px`,
            }}
            src={`${srcThumbnail}`}
          />
        </div>
        {showScrub && (
          <div
            className="absolute h-full w-[2px] bg-gray-200"
            style={{
              left: scrubPosition,
            }}
          />
        )}
      </div>
    </div>
  )
}

export default MediaCardThumbnail
