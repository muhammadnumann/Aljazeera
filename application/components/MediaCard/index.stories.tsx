import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import Component from '.'
import { getFakeMetaCardItemType } from 'stubs/ItemMetadata.stubs'

export default {
  title: 'Components/MediaCard',
  component: Component,
  argTypes: {},
} as ComponentMeta<typeof Component>

const Template: ComponentStory<typeof Component> = (args) => (
  <Component {...args} />
)

export const Default = Template.bind({})

Default.args = {
  itemType: getFakeMetaCardItemType(),
  parseMetadataOptions: { flat: true, arrayOnSingle: false },
}

Default.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/z3UqTsgsPM9SS2793hfBEJ/AJD-MAM-UI?node-id=378%3A10145',
  },
}
