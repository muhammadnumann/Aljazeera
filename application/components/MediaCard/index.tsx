import {
  useMemo,
  MouseEvent,
  useCallback,
  useState,
  useEffect,
  useContext,
} from 'react'
import { parseMetadataType } from '@vidispine/vdt-js'
import { getMetadataTitle } from 'utils/MetadataUtils'
import MediaCardThumbnail from './MediaCardThumbnail'
import MediaCardContent from './MediaCardContent'
import MediaCardMoreDetails from './MediaCardMoreDetails'
import { ItemType } from '@vidispine/types'
import TimeAgo from 'timeago-react'
import { Api } from 'utils/Api'
import { getItemThumbnailSpriteSheet } from 'utils/vdt-api/item'
import AppContext from 'utils/contexts/AppContext'
const { XMLParser } = require('fast-xml-parser')

interface InputMetadataProp {
  flat: boolean
  arrayOnSingle: boolean
}

interface InputProps {
  itemType: ItemType
  onClick(itemType: ItemType, event: MouseEvent<HTMLDivElement>): void
  onRemoveItemClick: () => void
  parseMetadataOptions: InputMetadataProp
  primaryFieldName: () => string
  primaryFunction: ({ item_videoId, itemId }: MediaCardProp) => void
}

const MediaCard = (props: InputProps) => {
  const context = useContext(AppContext)
  const parser = useMemo(() => new XMLParser(), [])
  const { id, metadata: metadataType } = props.itemType
  const [spriteImage, setSpriteImage] =
    useState<HTMLImageElement | undefined>(undefined)
  const [arraySpriteImage, setArraySpriteImage] = useState([])
  const parsedMetadata = parseMetadataType(
    metadataType,
    props.parseMetadataOptions,
  )
  const primaryText = useMemo(
    () =>
      getMetadataTitle(
        props.primaryFieldName,
        props.primaryFunction,
        parsedMetadata,
        props.itemType,
        id,
      ) || id,
    [
      props.primaryFieldName,
      props.primaryFunction,
      parsedMetadata,
      props.itemType,
      id,
    ],
  )

  const getThumbnailSprite = useCallback(async () => {
    const response = await Api.fetch(getItemThumbnailSpriteSheet, {
      itemId: id,
    })
    const data = parser.parse(response.data)
    if (data.ThumbnailSpriteSheetDocument) {
      const sprite = new Image()
      sprite.src = data.ThumbnailSpriteSheetDocument.url
      sprite.onload = () => {
        setSpriteImage(sprite)
        setArraySpriteImage(data.ThumbnailSpriteSheetDocument.thumbnail)
      }
    }
  }, [id, parser])

  useEffect(() => {
    getThumbnailSprite()
  }, [id, getThumbnailSprite])

  const { item_airDate: itemAirDate, representativeThumbnailNoAuth } =
    parsedMetadata

  const bgImage = useMemo(() => {
    if (representativeThumbnailNoAuth) {
      return context.serverUrl + representativeThumbnailNoAuth
    }
    return ''
  }, [representativeThumbnailNoAuth, context.serverUrl])

  return (
    <div className="grid h-60 w-64 rounded-lg shadow-lg">
      <div
        className={'h-[140px] w-64 rounded-t-lg '}
        style={{ backgroundImage: `url(${bgImage})` }}
      >
        <MediaCardThumbnail
          thumbnail={representativeThumbnailNoAuth}
          itemId={id}
          onClick={(event: MouseEvent<HTMLDivElement>) =>
            props.onClick(props.itemType, event)
          }
          spriteImage={spriteImage}
          thumbnailArray={arraySpriteImage}
        />
      </div>
      <div className="relative flex h-[100px] w-64 bg-gray-50 dark:bg-grayDark-50 rounded-b-lg border-t-[4px] border-t-orange-500">
        <MediaCardContent
          content={primaryText}
          slug={parsedMetadata.item_slug}
          className="self-center pl-4 -mt-4 text-[14px] text-gray-800 leading-[20px] font-semibold dark:text-white cursor-pointer"
          onClick={(event: MouseEvent<HTMLDivElement>) =>
            props.onClick(props.itemType, event)
          }
        />
        <TimeAgo
          className="absolute left-4 bottom-2 text-[14px] text-gray-700 dark:text-grayDark-400"
          datetime={itemAirDate}
        />
        <MediaCardMoreDetails
          content={primaryText}
          itemType={props.itemType}
          onRemoveItemClick={props.onRemoveItemClick}
        />
      </div>
    </div>
  )
}

export interface MediaCardProp {
  title: string
  itemId: string
  item_videoId: string
}

MediaCard.defaultProps = {
  itemType: {},
  primaryFieldName: undefined,
  primaryFunction: ({ item_videoId, itemId }: MediaCardProp) =>
    item_videoId || itemId,
  onClick: undefined,
  parseMetadataOptions: { flat: true, arrayOnSingle: false },
}

export default MediaCard
