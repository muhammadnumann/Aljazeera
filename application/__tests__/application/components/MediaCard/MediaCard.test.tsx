import React from 'react'
import MediaCard from '../../../../components/MediaCard/index'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'

describe('<MediaCard />', () => {

    it('should render correctly', () => {
        const component = shallow(<MediaCard />);
        expect(toJson(component)).toMatchSnapshot();
    });
});
