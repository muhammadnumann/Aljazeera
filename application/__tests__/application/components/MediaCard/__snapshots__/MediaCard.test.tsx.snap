// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaCard /> should render correctly 1`] = `
<div
  className="grid h-60 w-64 rounded-lg shadow-lg"
>
  <div
    className="h-[140px] w-64 rounded-t-lg "
    style={
      {
        "backgroundImage": "url()",
      }
    }
  >
    <MediaCardThumbnail
      onClick={[Function]}
      thumbnailArray={[]}
    />
  </div>
  <div
    className="relative flex h-[100px] w-64 bg-gray-50 dark:bg-grayDark-50 rounded-b-lg border-t-[4px] border-t-orange-500"
  >
    <MediaCardContent
      className="self-center pl-4 -mt-4 text-[14px] text-gray-800 leading-[20px] font-semibold dark:text-white cursor-pointer"
      onClick={[Function]}
      truncateContentLength={30}
    />
    <TimeAgo
      className="absolute left-4 bottom-2 text-[14px] text-gray-700 dark:text-grayDark-400"
      live={true}
    />
    <MediaCardMoreDetails
      itemType={{}}
    />
  </div>
</div>
`;
